### PR TITLE
Support GET/PUT with remote address out of segment, minor cleanups

### PR DESCRIPTION
**Pull request by mppf**
* Makes chpl_comm_get and chpl_comm_put check to see if the remote address is
  in the comms segment. If not, it uses an active message to complete the
  operation. For a GET, it sends an active message requesting a reply with the
  data. For a PUT, it sends the data in an active message and the handler
  memcpy's the data to the proper location.
* Generalized the AckArg0 and AckArg1 functionality to pack a pointer into two
  32-bit arguments, since I needed to use that to implement the new
  functionality.
* Move recurring sequence to wait for a reply to an active message into
  do_wait_done and WAIT_DONE_OBJ since I needed to use it too here.

This patch only adds this support to chpl_comm_get and chpl_comm_put.  I'd like
advise on how/whether or not to support this for the strided put/get calls or
for the nonblocking variants.

### CODE REVIEW

**Comment by mppf** [comment]
@gbtitus could you look at this? Any advise on the strided calls? Thanks.

File: runtime/src/comm/gasnet/comm-gasnet.c [hunk 1]
```diff
@@ -49,14 +50,37 @@ static chpl_commDiagnostics chpl_comm_commDiagnostics;
 static int chpl_comm_no_debug_private = 0;
 static gasnet_seginfo_t* seginfo_table = NULL;
 
+// Gasnet AM handler arguments are only 32 bits, so here we have
+// functions to get the 2 arguments for a 64-bit pointer,
+// and a function to reconstitute the poiter from the 2 arguments.
```
Comment (gbtitus) [inline runtime/src/comm/gasnet/comm-gasnet.c]: Typo: "poiter".

File: runtime/src/comm/gasnet/comm-gasnet.c [hunk 2]
```diff
@@ -255,6 +304,9 @@ static void fork_nb_large_wrapper(fork_t* f) {
   void* f_arg;
   chpl_memcpy(&f_arg, f->arg, sizeof(void*));
 
+  // MPF - should internal to the GASNet comms layer communication
+  // like this be counted in e.g. chpl_comm_commDiagnostics.get ?
+  // If not, we probably want to make e.g. do_remote_get.
```
Comment (gbtitus) [inline runtime/src/comm/gasnet/comm-gasnet.c]: (Just a comment.) Historically we've counted these. Arguably as especially the module code gets more and more abstract these counters may come to have less and less meaning at least to a user programmer, but for now it's probably best to just count everything, no matter who initiates it. I'm not sure that putting the question in an obscure comment deep in the comm layer is the right place, though. Might be better to ask on one of the mailing lists!

File: runtime/src/comm/gasnet/comm-gasnet.c [hunk 3]
```diff
@@ -340,6 +390,32 @@ static void AM_bcast_seginfo(gasnet_token_t token, void *buf, size_t nbytes) {
   bcast_seginfo_done = 1;
 }
 
+// Put from arg->src (which is local to the AM handler) back to
+// arg->dst (which is local to the caller of this AM).
+// nbytes must be < gasnet_AMMaxLongReply.
```
Comment (gbtitus) [inline runtime/src/comm/gasnet/comm-gasnet.c]: Is this condition checked anywhere?

File: runtime/src/comm/gasnet/comm-gasnet.c [hunk 4]
```diff
@@ -814,7 +893,58 @@ void  chpl_comm_put(void* addr, c_nodeid_t node, void* raddr,
       chpl_comm_commDiagnostics.put++;
       chpl_sync_unlock(&chpl_comm_diagnostics_sync);
     }
-    gasnet_put(node, raddr, addr, size); // node, dest, src, size
+
+    // Handle remote address not in remote segment.
+#ifdef GASNET_SEGMENT_EVERYTHING
```
Comment (gbtitus) [inline runtime/src/comm/gasnet/comm-gasnet.c]: The code would be cleaner if the #ifdef here were moved inside of chpl_comm_is_in_segment(). Currently that fn actually has such an #ifdef in it, with unfortunately exactly the opposite sense. Is there no way to reconcile these two? The combination of the behavior of the #ifdef outside chpl_comm_is_in_segment() and the identical but opposite-sensed one inside it seems bizarre.

File: runtime/src/comm/gasnet/comm-gasnet.c [hunk 5]
```diff
@@ -835,7 +967,97 @@ void  chpl_comm_get(void* addr, c_nodeid_t node, void* raddr,
...
+      // If the local address isn't in a registered segment,
+      // do the GET into a temporary buffer instead, and then
+      // copy the result back.
+      if( ! local_in_segment ) {
+        size_t buf_sz = size;
+        if( buf_sz > max_chunk ) {
+          buf_sz = max_chunk;
+        }
+
+        local_buf = chpl_mem_alloc(buf_sz, CHPL_RT_MD_COMM_XMIT_RECV_BUF, 0, 0);
```
Comment (gbtitus) [inline runtime/src/comm/gasnet/comm-gasnet.c]: A sanity check assert that local_buf is_in_segment might be a good idea here.

File: runtime/src/comm/gasnet/comm-gasnet.c [hunk 6]
```diff
@@ -835,7 +967,97 @@ void  chpl_comm_get(void* addr, c_nodeid_t node, void* raddr,
...
+      }
+
+      // do a PUT on the remote locale back to here.
+      // But do it in chunks of size gasnet_AMMaxLongReply()
+      // since we use A
```
Comment (gbtitus) [inline runtime/src/comm/gasnet/comm-gasnet.c]: Was there supposed to be more in this comment?

File: runtime/src/comm/gasnet/comm-gasnet.c [hunk 7]
```diff
@@ -1096,14 +1324,9 @@ void  chpl_comm_fork_fast(c_nodeid_t node, c_sublocid_t subloc,
...
-#ifndef CHPL_COMM_YIELD_TASK_WHILE_POLLING
-      GASNET_BLOCKUNTIL(done.flag);
-#else
-      while (!done.flag) {
-        (void) gasnet_AMPoll();
-        chpl_task_yield();
-      }
-#endif
+
+      WAIT_DONE_OBJ(done);
```
Comment (gbtitus) [inline runtime/src/comm/gasnet/comm-gasnet.c]: +1

Like this change a lot!

File: runtime/src/comm/gasnet/comm-gasnet.c [hunk 8]
```diff
@@ -118,6 +142,21 @@ typedef struct {
...
+  GASNET_BLOCKUNTIL(done->flag);
+#else
+  while (!done->flag) {
+    (void) gasnet_AMPoll();
+    chpl_task_yield();
+  }
+#endif
+}
+
+#define WAIT_DONE_OBJ(done) do_wait_done(&done)
```
Comment (gbtitus) [inline runtime/src/comm/gasnet/comm-gasnet.c]: For safety, should be do_wait_done(&(done)), in case the macro arg has low-precedence operators in it. Alternatively, just have the users of this invoke do_wait_done(&whatever) themselves!

**Comment by gbtitus** [comment]
General comment in response to the one in the commit: you probably can skip handling not-in-segment memory for the strided stuff for now. I expect most such calls will refer to heap space, so the fact that qthreads-based tasking gets its stacks from plain old malloc() won't affect those.
Beyond that, modulo my comments on the code, looks good and can be merged. I don't need to review it again if you make changes for the code comments.

File: runtime/src/comm/gasnet/comm-gasnet.c (hunk 2 above)
Comment (mppf) [inline runtime/src/comm/gasnet/comm-gasnet.c]: That's true, but I got you to answer it didn't I? :) I sometimes put comments like this that are really questions for my reviewer. I agree the comment doesn't have much value after this PR is merged.

File: runtime/src/comm/gasnet/comm-gasnet.c (hunk 8 above)
Comment (mppf) [inline runtime/src/comm/gasnet/comm-gasnet.c]: OK, I've switched WAIT_DONE_OBJ and INIT_DONE_OBJ to static inline functions.

File: runtime/src/comm/gasnet/comm-gasnet.c (hunk 1 above)
Comment (mppf) [inline runtime/src/comm/gasnet/comm-gasnet.c]: Fixed.

File: runtime/src/comm/gasnet/comm-gasnet.c (hunk 2 above)
Comment (mppf) [inline runtime/src/comm/gasnet/comm-gasnet.c]: Also, isn't this a source of test system noise, because different comms layers emit different numbers of GETs?

File: runtime/src/comm/gasnet/comm-gasnet.c (hunk 3 above)
Comment (mppf) [inline runtime/src/comm/gasnet/comm-gasnet.c]: It's not checked - it's just written to meet that requirement.  The comment around line 1023 is describing how the loop calling AM_reply_put (called DO_REPLY_PUT when it calls it) is using chunks of size <gasnet_AMMaxLongReply.
Would you like me to add an assert here?

File: runtime/src/comm/gasnet/comm-gasnet.c (hunk 4 above)
Comment (mppf) [inline runtime/src/comm/gasnet/comm-gasnet.c]: I agree that the result here is a bit strange. In order to fix it though don't we need to change the API in chpl-comm.h? The source of the issue is really differing ideas of what "is in segment" means.
* for the cache system, we need to know if a pointer could definitely be dereferenced on the remote locale without causing core dump, since we want to use the feature to do more memory requests.
* for this patch, we just need to know if we can PUT/GET to that address and have it work if that address is valid on the remote side.

So, maybe we should differentiate between these three cases:
* PUT/GET to that address would not succeed
* PUT/GET to that address would succeed if the address is valid (mapped, not NULL, etc)
* PUT/GET to that address would succeed and the address is known to be valid (in a registered heap)

What if we changed chpl_comm_is_in_segment to differentiate these three cases? I'd probably have
it return one of 3 values depending on the situation - they could be 0,1,2 or we could use an enum/#define them.

Thoughts? This would change other comms layers (but perhaps not in a way that matters since the cache currently only works for gasnet and I think that's the only current caller of chpl_comm_is_in_segment).

File: runtime/src/comm/gasnet/comm-gasnet.c (hunk 5 above)
Comment (mppf) [inline runtime/src/comm/gasnet/comm-gasnet.c]: OK, added.

**Comment by mppf** [comment]
@gbtitus, I've updated the patch with your comments. My main question is whether or not you think it's worth adjusting the API for chpl_comm_is_in_segment as I was describing in a comment - and if it's worth doing that in this patch. (I could commit this one and then do a separate PR cleaning this one up and making that change).

File: runtime/src/comm/gasnet/comm-gasnet.c (hunk 2 above)
Comment (gbtitus) [inline runtime/src/comm/gasnet/comm-gasnet.c]: Response #1: Yes, I answered it, but now my answer will be buried in the source control system instead of being more accessible, in the mailing list archives. (Neither is great, but the latter is definitely better.)

Response #2: It's a source of test system differences (from comm layer to comm layer) but these differences aren't noise, they're signal. They reflect implementation differences between the comm layer implementations and can help guide programmers to use the right comm layer for their particular application. So they're desirable differences, which we just have to deal with using any of variety of testing system techniques.

File: runtime/src/comm/gasnet/comm-gasnet.c (hunk 3 above)
Comment (gbtitus) [inline runtime/src/comm/gasnet/comm-gasnet.c]: Ah, I misread the comment as if it were expressing a requirement rather than a truth. Maybe rephrase as something like "nbytes is < gasnet_AMMaxLongReply here (see chpl_comm_get)."

File: runtime/src/comm/gasnet/comm-gasnet.c (hunk 4 above)
Comment (gbtitus) [inline runtime/src/comm/gasnet/comm-gasnet.c]: These are two different questions and we should have two different functions to answer them. Although it seems like there are 3 cases, I believe under some circumstances there are actually 4. For segment=fast or =everything, where the address is in the GASNet segment but is in a page that doesn't physically exist yet because it hasn't ever been referenced, it would be in-segment but not referenceable, at least for a read. (You could write to it and this would cause it to come into existence, but reading from it would produce a segfault. Stack autogrow regions are like this.) So, I would be in favor of  renaming the current function as, say, chpl_comm_addr_exists(), and naming the new one chpl_comm_is_in_segment().

However, I don't think it's necessary or desirable to mix that change up with this one. You could either do a separate PR to rename the existing function first, or (much easier) do this PR first, as is, and then split the functions apart afterward.

**Comment by mppf** [comment]
- [ ] GASNET+everything examples tests pass
- [ ] GASNET+fast examples tests pass
